singleheader/amalgamate.py

```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/python3
+
+import sys
+import os.path
+import subprocess
+from pathlib import Path
+import re
+
+SCRIPTPATH=os.path.dirname(os.path.abspath(sys.argv[0]))
+PROJECTPATH=os.path.dirname(SCRIPTPATH)
+
+print(f"SCRIPTPATH={SCRIPTPATH} PROJECTPATH={PROJECTPATH}")
+
+
+print("We are about to amalgamate all simdjson files into one source file.")
+print("See https://www.sqlite.org/amalgamation.html and https://en.wikipedia.org/wiki/Single_Compilation_Unit for rationale.")
+if "AMALGAMATE_SOURCE_PATH" not in os.environ:
+  AMALGAMATE_SOURCE_PATH=os.path.join(PROJECTPATH,"src")
+else:
+  AMALGAMATE_SOURCE_PATH=os.environ["AMALGAMATE_SOURCE_PATH"]
+if "AMALGAMATE_INCLUDE_PATH" not in os.environ:
+  AMALGAMATE_INCLUDE_PATH=os.path.join(PROJECTPATH,"include")
+else:
+  AMALGAMATE_INCLUDE_PATH=os.environ["AMALGAMATE_INCLUDE_PATH"]
+if "AMALGAMATE_OUTPUT_PATH" not in os.environ:
+  AMALGAMATE_OUTPUT_PATH=os.path.join(SCRIPTPATH)
+else:
+  AMALGAMATE_OUTPUT_PATH=os.environ["AMALGAMATE_OUTPUT_PATH"]
+
+# this list excludes the "src/generic headers"
+ALLCFILES=["simdjson.cpp"]
+
+# order matters
+ALLCHEADERS=["simdjson.h"]
+
+found_includes=[]
+
+
+def doinclude(fid,file,line):
+  p=Path(AMALGAMATE_INCLUDE_PATH)/file
+  pi=Path(AMALGAMATE_SOURCE_PATH)/file
+
+  if p.exists():
+    # generic includes are included multiple times
+    if re.match('.*generic/.*.h',file):
+      dofile(fid,AMALGAMATE_INCLUDE_PATH,file)
+    # begin/end_implementation are also included multiple times
+    elif re.match('.*/begin.h',file):
+      dofile(fid,AMALGAMATE_INCLUDE_PATH,file)
+    elif re.match('.*/end.h',file):
+      dofile(fid,AMALGAMATE_INCLUDE_PATH,file)
+    elif file not in found_includes:
+      found_includes.append(file)
+      dofile(fid,AMALGAMATE_INCLUDE_PATH,file)
+  elif pi.exists():
+    # generic includes are included multiple times
+    if re.match('.*generic/.*h',file):
+      dofile(fid,AMALGAMATE_SOURCE_PATH,file)
+    elif file not in found_includes:
+      found_includes.append(file)
+      dofile(fid,AMALGAMATE_SOURCE_PATH,file)
+    else:
+      print(f"//doinclude: {file} already included: {line}")
+  else:
+    # If we don't recognize it, just emit the #include
+    print(line,file=fid)
+
+def dofile(fid,prepath,filename):
+  #print(f"// dofile: invoked with prepath={prepath}, filename={filename}",file=fid)
+  file=Path(prepath)/filename
+  RELFILE=file.relative_to(PROJECTPATH)
+  # Last lines are always ignored. Files should end by an empty lines.
+  print(f"/* begin file {RELFILE} */",file=fid)
+  fid2=open(file,'r')
```

Comment:
This is leaking file handles, you never close fid2. I'd suggest using it as a context manager.

`with open(file, 'r') as fid2:`

Reply:
Thanks, fixed.